File: src/advection.jl
"""
$(SIGNATURES)

Return the value at x in [0,1] of the B-spline with integer nodes of degree
p with support starting at j.  Implemented recursively using the 
[De Boor's Algorithm](https://en.wikipedia.org/wiki/De_Boor%27s_algorithm)
```math
B_{i,0}(x) := \\left\\{
\\begin{matrix}
1 & \\mathrm{if}  \\quad t_i ≤ x < t_{i+1} \\\\
0 & \\mathrm{otherwise} 
\\end{matrix}
\\right.
```
```math
B_{i,p}(x) := \\frac{x - t_i}{t_{i+p} - t_i} B_{i,p-1}(x) 
+ \\frac{t_{i+p+1} - x}{t_{i+p+1} - t_{i+1}} B_{i+1,p-1}(x).
```
"""
function bspline(p::Int, j::Int, x::Float64)
    if p == 0
        j == 0 ? (return 1.0) : (return 0.0)
    else
        w = (x - j) / p
        w1 = (x - j - 1) / p
    end
    return w * bspline(p - 1, j, x) + (1 - w1) * bspline(p - 1, j + 1, x)
end

""" 
$(TYPEDEF)

Advection to be computed on each row

$(TYPEDFIELDS)
"""
struct Advection

    mesh::Mesh
    p::Int
    modes::Vector{Float64}
    eig_bspl::Vector{Float64}
    eigalpha::Vector{ComplexF64}

    function Advection(mesh, p = 3)
        modes = [2pi * i / mesh.nx for i = 0:mesh.nx-1]
        eig_bspl = zeros(mesh.nx)
        eig_bspl .= bspline(p, -div(p+1,2), 0.0)
        for i in 1:div(p+1,2)-1
            eig_bspl .+= bspline(p, i - (p+1)÷2, 0.0) * 2 .* cos.(i * modes)
        end
        eigalpha = zeros(ComplexF64, mesh.nx)
        new(mesh, p, modes, eig_bspl, eigalpha)
    end

end

export advect

"""
$(SIGNATURES)
"""
function advect(self, f, v, dt)

    p  = self.p
    nx = self.mesh.nx
    dx = self.mesh.dx

    ft = fft(f, 1)

    for j in eachindex(v)

        alpha = dt * v[j] / dx
        # compute eigenvalues of cubic splines evaluated at displaced points
        ishift = floor(-alpha)
        beta = -ishift - alpha
        fill!(self.eigalpha, 0.0im)
        for i = -div(p - 1, 2):div(p + 1, 2)
            self.eigalpha .+= (
                bspline(p, i - div(p + 1, 2), beta) .*
                exp.((ishift + i) .* 1im .* self.modes)
            )
        end

        # compute interpolating spline using fft and properties 
        # of circulant matrices

        ft[:, j] .*= self.eigalpha ./ self.eig_bspl

    end

    f .= real(ifft(ft, 1))

end

File: test/advection.py
import numpy as np
import multiprocessing as mp
import numpy as np

def bspline(p, j, x):
	"""
	Return the value at x in [0,1[ of the B-spline with
	integer nodes of degree p with support starting at j.
	Implemented recursively using the de Boor's recursion formula
	"""
	if p == 0:
		if j == 0:
			return 1.0
		else:
			return 0.0
	else:
		w = (x - j) / p
		w1 = (x - j - 1) / p
	return (w * bspline(p - 1, j, x) + (1 - w1) * bspline(p - 1, j + 1, x))


class Advection:
	""" Advection to be computed on each row """
	def __init__(self, mesh, p=3):
		self.mesh = mesh
		self.p     = p
		nx        = mesh.nx
		#modes    = np.zeros(nx, np.float64)
		self.modes    = np.array([2*np.pi * i / nx for i in range(0,nx)])
		self.eig_bspl = np.zeros(nx, np.float64)
		self.eig_bspl.fill(bspline(p, -np.floor((p+1)/2), 0.0))
		for i in range(1,int(np.floor(p+1/2))):
			self.eig_bspl += bspline(p, i-np.floor((p+1)/2), 0.0) * 2 * np.cos(i * self.modes)

		self.eigalpha  = np.zeros(nx, np.complex128)

	def advect_iteration(self, j):
		f_fft = self.f_fft
		v = self.v
		dt = self.dt
		nx = self.mesh.nx
		dx = self.mesh.dx
		alpha = dt * v[j] / dx
		# compute eigenvalues of cubic splines evaluated at displaced points
		ishift = np.floor(-alpha)
		beta   = -ishift - alpha
		eigalpha  = np.zeros(nx, np.complex128)

		for i in range(-int(np.floor((self.p - 1) / 2)), int(np.floor((self.p + 1) / 2) + 1)):
				eigalpha += (bspline(self.p, i - np.floor((self.p + 1) / 2), beta)
				* np.exp((ishift + i) * 1j * self.modes))

		f_fft[j, :] *= self.eigalpha / self.eig_bspl
		return (j, np.real(np.fft.ifft(f_fft[j, :] * eigalpha / self.eig_bspl)))

	def advect(self, f, v, dt):
		nx = self.mesh.nx
		nv = len(v)
		dx = self.mesh.dx

		f_fft = np.fft.fft(f, axis=1)

		#Parallel(n_jobs=1)(delayed(self.advect_iteration)(f_fft, v, j, dt) for j in range(0,nv))

		# with mp.Pool(processes=2) as pool:
		# 	self.f_fft = f_fft
		# 	self.v = v
		# 	self.dt = dt
		# 	result = pool.map(self.advect_iteration, range(nv))
		# for row in result:
		# 	i = row[0]
		# 	f[i, :] = row[1]

		for j in range(0, nv):
			alpha = dt * v[j] / dx
			# compute eigenvalues of cubic splines evaluated at displaced points
			ishift = np.floor(-alpha)
			beta   = -ishift - alpha
			self.eigalpha.fill(0.0j)
			for i in range(-int(np.floor((self.p - 1) / 2)), int(np.floor((self.p + 1) / 2) + 1)):
				self.eigalpha += (bspline(self.p, i - np.floor((self.p + 1) / 2), beta)
					* np.exp((ishift + i) * 1j * self.modes))
			f_fft[j, :] *= self.eigalpha / self.eig_bspl

		f[:, :] = np.real(np.fft.ifft(f_fft, axis=1))


